Picard output

permcody opened the pull request:
<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

Several enhancements to the format of Picard outputs. Here's a brief sample without color:
```

Beginning Picard Iteration 4

Starting Transfers on TIMESTEP_BEGIN To MultiApps
Beginning NearestNodeTransfer u_to_sub
Finished NearestNodeTransfer u_to_sub
Jobs complete: 1/1
Waiting For Transfers To Finish
Transfers on TIMESTEP_BEGIN Are Finished


Executing MultiApps on TIMESTEP_BEGIN
Solving MultiApp sub
Solving Normal Step!
sub0: 
sub0: Time Step 10, time = 1
sub0:                 dt = 0.1
sub0: 
sub0:  0 Nonlinear |R| = 2.787993e-09
sub0:       0 Linear |R| = 2.787993e-09
sub0:       1 Linear |R| = 3.423644e-10
sub0:       2 Linear |R| = 8.127254e-12
sub0:       3 Linear |R| = 1.095456e-13
sub0:       4 Linear |R| = 2.059790e-15
sub0:  1 Nonlinear |R| = 2.619825e-15
sub0:  Solve Converged!
Successfully Solved MultiApp sub.
Jobs complete: 1/1
Waiting For Other Processors To Finish
Finished Executing MultiApps on TIMESTEP_BEGIN


Starting Transfers on TIMESTEP_BEGIN From MultiApps
Beginning NearestNodeTransfer v_from_sub
Finished NearestNodeTransfer v_from_sub
Jobs complete: 1/1
Waiting For Transfers To Finish
Transfers on TIMESTEP_BEGIN Are Finished

Picard Norm after TIMESTEP_BEGIN MultiApps: 2.801717e-10

Time Step 10, time = 1
                dt = 0.1

 0 Nonlinear |R| = 2.801717e-10
      0 Linear |R| = 2.801717e-10
      1 Linear |R| = 5.937284e-12
      2 Linear |R| = 8.103817e-15
      3 Linear |R| = 5.098351e-17
 1 Nonlinear |R| = 1.540961e-16
 Solve Converged!

 0 Picard |R| = 4.026807e-02
 1 Picard |R| = 3.670046e-04
 2 Picard |R| = 3.353973e-06
 3 Picard |R| = 3.065428e-08
Picard converged!
```

permcody committed:
Update Picard output

1. Print Picard convergence history
2. Remove unneeded Picard norms
3. Add more color Picard norms

closes #12708

permcody committed:
New Picard output tests

refs #12708

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/276551/) on 0d7ba63 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/12709/site)

This comment will be updated on new commits.

YaqiWang commented:
I really like this move! One question, can this change close #12255?

permcody commented:
FYI - Sabertooth is pretty hosed. I really need this PR merged before I want to bother fixing it. It's technically not my fault, it was broken with https://github.com/idaholab/moose/commit/62fb4cb026a12ad4d0527f180ab6db63a81fbffe

YaqiWang commented:
I was wondering what happened to Sabertooth. That commit seems logically correct. What happened? `TransientMultiApp` needs to be cleaned up. It has much more code duplication. Typically this kind of small issues, that is hard to fix, indicates problematic designs in the code. We assume timestep_end is always picard_end, which may not always be desired to the applications. So it is not an good argument of Picard iteration being outside of timestep.

permcody commented:
Vincent changed (fixed) the behavior of what MOOSE was doing when it hit "Picard Max Its". Previously MOOSE was accepting (which was wrong), now it's rejecting, so that test now times out because it takes way longer to converge with the rejected time step. It needs either a tolerance tweak or smaller problem, etc. I'll fix it, but I'd like this PR in first.

permcody commented:
@YaqiWang - Any idea what's going on in the Rattlesnake tutorials? Perhaps removing the TIMESTEP_END norms changed something here, although I don't immediately see why...

vincentlaboure commented:
@permcody We are thinking of adding an extra option/parameter like @friedmud suggested (#12641) which would allow to move on to the next time step if the maximum number of iterations is reached.

While this is not often useful, there are cases where it can be, such as pseudo-transients, where the first time steps can struggle reducing the picard norm efficiently. Simply using relative and absolute tolerances may not be enough or extremely inefficient.

In that case, we could easily fix the Sabertooth failing case by replacing `picard_max_its` with that new parameter.

YaqiWang commented on framework/src/executioners/Transient.C:
```diff
@@ -531,12 +540,15 @@ Transient::solveStep(Real input_dt)
   _problem.execute(EXEC_TIMESTEP_BEGIN);
 
-  if (_picard_max_its > 1)
+  if (_picard_max_its > 1 && _problem.hasMultiApps(EXEC_TIMESTEP_BEGIN))
```
Potentially any aux kernel or user object of master on timestep_begin can change the residual evaluation of master. Unless we can find a rigorous check, we should drop this `&&` here unfortunately. We can provide an advanced input parameter `skip_timestep_begin_picard_norm_evaluation`.

YaqiWang replied:
Another thinking is that when we add kernels, dgkernels or any objects for residual evaluation, if it couples aux variables or user objects, we can check mark a flag to indicate the kernel evaluation is affected by execution of timestep_begin, timestep_end, etc. Then based on this flag, we can bypass the norm evaluation here. Not sure if it is worthwhile to do this.

YaqiWang replied:
Sorry, this is not going to work if we have something like this: kernel => material => auxkernel.

permcody replied:
I added a flag that will allow you to restore the existing behavior. Will that work for you? If I get a chance, I'll see if that fixes the tutorial failures in Rattlesnake.

permcody committed:
Add picard_force_norms flag

Allow a developer to choose whether they would like to
force both a begin and end evaluation of the norm
to account for things like coupled value lag that would
impact the norm.

permcody commented:
@YaqiWang - I added a flag and have a patch for Rattlesnake. Using the flag maintains the current gold standard.

YaqiWang commented:
Thanks. I tend to make that parameter default to true. But if your guys insist, I am ok.

friedmud reviewed (approved):
👍

YaqiWang commented:
Sorry @permcody . There are several heavy tests in Rattlesnake also impacted by this change: https://moosebuild.inl.gov/job/154531/.
